python/mujoco_mpc/demos/direct_optimizer.py

```diff
@@ -0,0 +1,1479 @@
+    weights_sensor: weights for sensor norms (nsensor x horizon).
+    weights_force: weights for force norms (nv x horizon).
+    _dvdq0: Jacobian of velocity wrt previous configuration ((nv x nv) x horizon).
+    _dvdq1: Jacobian of velocity wrt current configuration ((nv x nv) x horizon).
+    _dadq0: Jacobian of acceleration wrt previous configuration ((nv x nv) x horizon).
+    _dadq1: Jacobian of acceleration wrt current configuration ((nv x nv) x horizon).
+    _dadq2: Jacobian of acceleration wrt next configuration ((nv x nv) x horizon).
+    _dsdq: Jacobian of sensor wrt configuration ((nsensordata x nv) x horizon).
+    _dsdv: Jacobian of sensor wrt velocity ((nsensordata x nv) x horizon).
+    _dsda: Jacobian of sensor wrt acceleration ((nsensordata x nv) x horizon).
+    _dfdq: Jacobian of force wrt configuration ((nv x nv) x horizon).
+    _dfdv: Jacobian of force wrt velocity ((nv x nv) x horizon).
+    _dfda: Jacobian of force wrt acceleration ((nv x nv) x horizon).
+    _dsdq012: Jacobian of sensor wrt previous, current, and next configurations ((nv x 3 nv) x horizon).
+    _dfdq012: Jacobian of force wrt previous, current, and next configurations ((nv x 3 nv) x horizon).
+    cost_total: combined force and sensor costs.
+    cost_force: sum of weighted force norms.
+    cost_sensor: sum of weighted sensor norms.
+    cost_initial: initial total cost.
+    _ntotal: number of decision variables (nv * horizon).
+    _nband: cost Hessian band dimension (3 * nv).
+    _gradient: gradient of cost wrt to decision variables (nv * horizon).
+    _hessian: band representation of cost Hessian wrt decision variables (nv * horizon x 3 * nv).
+    _hessian_factor: factorization of band represented cost Hessian
+    _search_direction: Gauss-Newton search direction (nv * horizon).
+    _qpos_candidate: candidate search point for configuration trajectory (nq x horizon).
+    _regularization: current value for cost Hessian regularization.
+    _gradient_norm: normalized L2-norm of cost gradient.
+    _iterations_step: number of step iterations performed.
+    _iterations_search: number of curve search iterations performed.
+    _status_msg: status message.
+    gradient_tolerance: setting for solve termination based on cost gradient.
+    direction_tolerance: setting for solver termination based on search direction infinity norm.
+    max_iterations: setting for maximum number of step iterations performed.
+    max_search_iterations: setting for maximum number of search iterations performed at each step iteration.
+    regularization_min: minimum regularization value.
+    regularization_max: maximum regularization value.
+    regularization_scale: value for increasing/decreasing regularization.
+  """
+
+  def __init__(self, model: mujoco.MjModel, horizon: int):
+    # model + data
+    self.model = model
+    #TODO(taylor): set discrete inverse dynamics!!
+    # self.model.opt.enableflags |= self.model.mjtEnableBit.mjENBL_INVDISCRETE
```

Comment:
@yuvaltassa is this the correct way to set discrete inverse dynamics? I can't test without building python mujoco from source until the next release...